torch/_refs/nn/functional/__init__.py

```diff
@@ -294,3 +295,28 @@ def tanhshrink(a: TensorLikeType) -> TensorLikeType:
             "Expected a tensor input for an elementwise unary operation!"
         )
     return refs.sub(a, refs.tanh(a))
+
+
+@elementwise_unary_scalar_wrapper
+@elementwise_type_promotion_wrapper(
+    type_promoting_args=("a"),
+    type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT,
+)
+def hardtanh(
+    a: TensorLikeType, min_val: float = -1, max_val: float = 1, inplace: bool = False
+) -> TensorLikeType:
+    """
+    Reference implementation of torch.nn.functional.hardtanh
+    """
+    if inplace:
+        raise NotImplementedError
+    if utils.is_boolean_dtype(a.dtype):
+        raise RuntimeError("Bool inputs not supported for hardtanh")
+    if utils.is_integer_dtype(a.dtype):
+        min_val = int(min_val)
```

Comment:
This is going to require a comment -- so can min_val and max_val be float or int?

Reply:
Yeah, I think they should actually be `NumberLikeType`.

The comment in the ATen source is `    # preserve legacy behavior of boundaries not causing type promotion` - porting it over.